rfd/0abc-kube-access-forward-identity.md

```diff
@@ -0,0 +1,205 @@
+the behalf of the user. Since the proxy is authenticated using its certificate,
+the upstream service can trust the identity received in the request headers.
+
+This change will remove the requirement for the proxy to generate a new certificate
+and thus call auth's `ProcessKubeCSR` endpoint to sign the key-pair containing the
+user's identity.
+
+
+### Impersonation: Forward user identity using HTTP headers
+
+As opposed to the majority of protocols supported by Teleport, the Kubernetes protocol
+is based on HTTP 1/2. This allows us to forward the user's identity using HTTP headers
+for the specific case of Kubernetes.
+
+The proposed solution is to use the Proxy certificate to authenticate to the upstream
+service and to forward the user's identity using HTTP headers or gRPC metadata
+for later Impersonation.
+
+Once the upstream service receives a request, it will check the certificate provided
+to validate the request's provenance and its role - making sure it's a proxy - for
+authenticating the request. Once authenticated, it will impersonate the identity
+contained in the request headers. This impersonated identity will be available in
+`authz.Context` and services won't be able to distinguish between a request forwarded
+by the proxy or a request made directly by the user.
+
+This approach is similar to what Kubernetes API does when
+forwarding requests to the API server using Impersonation. In the Teleport case, we must
+forward the full user's identity instead of just the username/roles.
+
+Since the proxy has full access to the HTTP request, it can add the user's identity
+to the request headers before forwarding it to the upstream service.
+
+```go
+headers["TELEPORT_IMPERSONATE_IDENTITY"] = clientCert.Subject
+```
+
+In order to prevent the user from tampering with the headers, the authorization
+layer will delete the headers after checking if the request originated in a proxy,
+so the user cannot send the headers directly and the proxy forwards them
+to the upstream service.
+
+Besides the security benefits, this option also has the advantage of allowing us to
+reuse the same connection to forward multiple requests from different users. It happens
+because the HTTP headers are sent with each request and we do not require a new
+connection per request such as in the Proxy Protocol option. This will improve
+performance by reducing the number of connections that need to be established and
+the time it takes to forward a new request upstream. Currently, the proxy creates a new
+`http.Transport` per request which means that a new connection is established
+every time a request is forwarded. This is not ideal because it increases the
+latency of the request and it also increases the load on the upstream service.
+
+This option requires sending the user's IP address to the upstream service
+using a header instead of the protocol extension implemented by [TLS IP Pinning RFD](https://github.com/gravitational/teleport/pull/22481).
```

Comment:
Can it be just encoded as part of the identity in the impersonation header? I think source IP is already part of identity object.